### ISSUE TITLE
Coupon is applied properly on cart but not on checkout 

### ISSUE THREAD

**Issue by ioweb-gr** [magento/magento2#38746]
### Preconditions and environment

- 2.4.6-p4
- Luma theme
- Default checkout

### Steps to reproduce

1. Create a cart rule where the user can add two products from different manufactures but is supposed to receive a discount only on one 
2. The rule uses a COUPON CODE
3. Example conditions 
![image](https://github.com/magento/magento2/assets/20220341/66bb6ce5-699f-498f-b129-863e79420a53)
4. Example actions
![image](https://github.com/magento/magento2/assets/20220341/280ad8e3-a760-4096-84ca-5411f80b43fb)
5. Try to apply the coupon on cart and it works
![image](https://github.com/magento/magento2/assets/20220341/24705818-592e-4a49-a629-4de8085a17aa)
6. Try to go to checkout and apply it
![image](https://github.com/magento/magento2/assets/20220341/9a27a5e9-5634-40e3-bc5d-2759761440e4)

### Expected result

The coupon is properly applied

### Actual result

The coupon is not applied 

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by ioweb-gr** [comment]
@magento give me 2.4.6-p4 instance

**Comment by engcom-Dash** [comment]
Hi @ioweb-gr 

Thanks for reporting and collaboration.

Verified the issue on magento 2.4 dev instance but the issue is not reproducable.

The coupon is applied succesfully on both cart and checkout.

Steps to reproduce:

1. Create a cart rule where the user can add two products from different manufactures but is supposed to receive a discount only on one
2. The rule uses a COUPON CODE
3. Add conditions and actions.
4. On the UI, add the product with Manufacture with couponcode
5. Apply coupon code on cart
6. Apply coupon code on checkout

Please refer the screenrecording.

Uploading 38746-1 (1).mov…

**Comment by engcom-Dash** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Dash** [comment]
Hi @ioweb-gr 

Verified the issue again in magneto 2.4 dev instance. But the issue is not reproducable.

Able to apply the coupon code in both cart and checkout page

The rule has manufacturer and product belonging to a specific category.

Please refer the attached screenshots and recording.


https://github.com/magento/magento2/assets/60198592/dae32933-5328-4d9c-8d08-3f5c81df9b40

**Comment by ioweb-gr** [comment]
While trying to debug this I notice it fails here 

```

    /**
     * Validate coupon code.
     *
     * @param \Magento\Quote\Model\Quote $quote
     * @return $this
     */
    protected function _validateCouponCode(\Magento\Quote\Model\Quote $quote)
    {
        $code = $quote->getData('coupon_code');
        if ($code !== null && strlen($code)) {
            $addressHasCoupon = false;
            $addresses = $quote->getAllAddresses();
            if (count($addresses) > 0) {
                foreach ($addresses as $address) {
                    if ($address->hasCouponCode()) {
                        $addressHasCoupon = true;
                    }
                }
                if (!$addressHasCoupon) {
                    $quote->setCouponCode('');
                }
            }
        }
        return $this;
    }
```

While validating the code it's trying to check both shipping / billing address to verify if the coupon code is set. However, the coupon code is only set in the quote earlier 

`\Magento\Quote\Model\CouponManagement::set`

```
        try {
            $quote->setCouponCode($couponCode);
            $this->quoteRepository->save($quote->collectTotals());
        } catch (LocalizedException $e) {
            throw new CouldNotSaveException(__('The coupon code couldn\'t be applied: ' .$e->getMessage()), $e);
        } catch (\Exception $e) {
            throw new CouldNotSaveException(
                __("The coupon code couldn't be applied. Verify the coupon code and try again."),
                $e
            );
        }
```

So basically I'm assuming that the `$quote->setCouponCode($couponCode);` should also trigger setting the coupon code into the address itself. But it doesn't 

Could you point me to where the coupon code should be set in the address to see why it's not being set?

**Comment by ioweb-gr** [comment]
@magento give me 2.4.7 instance

**Comment by ioweb-gr** [comment]
@magento give me 2.4-develop instance

**Comment by ioweb-gr** [comment]
OK we found the culprit. There was a free shipping rule above X amount. When adding the two products, the free shipping rule was activating and preventing subsequent rules processing. Moreover, this was not enforced on cart but only on checkout hence the confusion. 

After upgrading to 2.4.7 to test our instance, this difference was eliminated and our coupon was not applying in both cart and checkout which was correct due to the settings on the free shipping rule. 

So there is an actual problem in 2.4.6-p4 where this use case causes mismatch but it's solved in 2.4.7. 

We can safely close this issue
